OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs

```diff
@@ -392,85 +392,94 @@ public bool CanQueue(ActorInfo actor, out string notificationAudio, out string n
+						return;
 
-						if (fromLimit <= 0)
-							return;
+					var isBuilding = unit.HasTraitInfo<BuildingInfo>();
+					if (isBuilding && !hasPlayedSound)
+					{
+						hasPlayedSound = Game.Sound.PlayNotification(rules, self.Owner, "Speech", Info.ReadyAudio, self.Owner.Faction.InternalName);
+						TextNotificationsManager.AddTransientLine(Info.ReadyTextNotification, self.Owner);
 					}
-
-					var cost = GetProductionCost(unit);
-					var time = GetBuildTime(unit, bi);
-					var amountToBuild = Math.Min(fromLimit, order.ExtraData);
-					for (var n = 0; n < amountToBuild; n++)
+					else if (!isBuilding)
 					{
-						var hasPlayedSound = false;
-						BeginProduction(new ProductionItem(this, order.TargetString, cost, playerPower, () => self.World.AddFrameEndTask(_ =>
+						if (BuildUnit(unit))
 						{
-							// Make sure the item hasn't been invalidated between the ProductionItem ticking and this FrameEndTask running
-							if (!Queue.Any(i => i.Done && i.Item == unit.Name))
-								return;
-
-							var isBuilding = unit.HasTraitInfo<BuildingInfo>();
-							if (isBuilding && !hasPlayedSound)
-							{
-								hasPlayedSound = Game.Sound.PlayNotification(rules, self.Owner, "Speech", Info.ReadyAudio, self.Owner.Faction.InternalName);
-								TextNotificationsManager.AddTransientLine(Info.ReadyTextNotification, self.Owner);
-							}
-							else if (!isBuilding)
-							{
-								if (BuildUnit(unit))
-								{
-									Game.Sound.PlayNotification(rules, self.Owner, "Speech", Info.ReadyAudio, self.Owner.Faction.InternalName);
-									TextNotificationsManager.AddTransientLine(Info.ReadyTextNotification, self.Owner);
-								}
-								else if (!hasPlayedSound && time > 0)
-								{
-									hasPlayedSound = Game.Sound.PlayNotification(rules, self.Owner, "Speech", Info.BlockedAudio, self.Owner.Faction.InternalName);
-									TextNotificationsManager.AddTransientLine(Info.BlockedTextNotification, self.Owner);
-								}
-							}
-						})), !order.Queued);
+							Game.Sound.PlayNotification(rules, self.Owner, "Speech", Info.ReadyAudio, self.Owner.Faction.InternalName);
+							TextNotificationsManager.AddTransientLine(Info.ReadyTextNotification, self.Owner);
+						}
+						else if (!hasPlayedSound && time > 0)
+						{
+							hasPlayedSound = Game.Sound.PlayNotification(rules, self.Owner, "Speech", Info.BlockedAudio, self.Owner.Faction.InternalName);
+							TextNotificationsManager.AddTransientLine(Info.BlockedTextNotification, self.Owner);
+						}
 					}
+				})), !order.Queued);
+			}
+		}
 
+		public void ResolveOrder(Actor self, Order order)
+		{
+			if (!Enabled)
+				return;
+
+			switch (order.OrderString)
+			{
+				case "StartProduction":
+					StartProduction(self, order);
+					break;
+				case "StartProductionAI":
+					// Hack: AI production modules won't produce on queue that is producing
+					// to avoid duplicated production due to order latency or network lags.
+					if (!AllQueued().Any())
```

Comment:
Why don't you move this into `UnitBuilderBotModule` and avoid the AI related code here with the strange custom order string?